### PR TITLE
NTS Hint for Eltwsise kernels

### PR DESCRIPTION
**Pull request by alheinecke**

### CODE REVIEW

File: src/generator_mateltwise_sse_avx_avx512.c [hunk 1]
```diff
@@ -302,6 +302,37 @@ void libxsmm_generator_mateltwise_footer_n_dyn_loop( libxsmm_generated_code*
...
+
+  if ( i_mateltwise_desc->operation == LIBXSMM_MELTW_OPERATION_UNARY ) {
+    if ( (i_mateltwise_desc->m   % (i_vlen_bytes / i_dt_width ) == 0) &&
+         (i_mateltwise_desc->ldo % (i_vlen_bytes / i_dt_width ) == 0) &&
+         ((i_mateltwise_desc->flags & LIBXSMM_MELTW_FLAG_UNARY_NTS_HINT) == LIBXSMM_MELTW_FLAG_UNARY_NTS_HINT ) ) {
+      l_instr = i_instr_nts;
+    } else {
+      l_instr = i_instr_st;
+    }
+  } else if ( i_mateltwise_desc->operation == LIBXSMM_MELTW_OPERATION_BINARY ) {
```
Comment (egeor) [inline src/generator_mateltwise_sse_avx_avx512.c:321]: Nit: we can merge the two ifs with logic (is_unary && unary_flag) || (is_binary && binary_flag) to reduce code duplication

**Review by egeor — APPROVED** [review]
Just minor cosmetic nits. If the plan is to use this hint flag within equations potentially we would need that there too?

File: samples/eltwise/eltwise_binary_simple.c [hunk 2]
```diff
@@ -536,6 +536,8 @@ int test_binary_op( const libxsmm_blasint M, const libxsmm_blasint N, const libx
     }
   }
 
+  binary_flags |= LIBXSMM_MELTW_FLAG_BINARY_NTS_HINT;
+
```
Comment (egeor) [inline samples/eltwise/eltwise_binary_simple.c:540]: So we want to test always with the hint flag?

**Review by egeor — APPROVED** [review]

File: samples/eltwise/eltwise_binary_simple.c (hunk 2 above)
Comment (alheinecke) [reply samples/eltwise/eltwise_binary_simple.c:540]: Reasoning: It's a hint let's turn it on in CI. Only very little case will execute it, but we will execute it. I think its not worth it to add 2x the test scripts.

File: include/libxsmm_typedefs.h [hunk 3]
```diff
@@ -242,9 +242,10 @@ typedef enum libxsmm_meltw_unary_flags {
   LIBXSMM_MELTW_FLAG_UNARY_REDUCE_NO_PREFETCH = 1024,
   LIBXSMM_MELTW_FLAG_UNARY_REDUCE_RECORD_ARGOP= 2048,
   LIBXSMM_MELTW_FLAG_UNARY_STOCHASTIC_ROUND   = 4096,
-  LIBXSMM_MELTW_FLAG_UNARY_GS_ROWS            = 8192,
-  LIBXSMM_MELTW_FLAG_UNARY_GS_COLS            = 16384,
-  LIBXSMM_MELTW_FLAG_UNARY_GS_OFFS            = 32768
+  LIBXSMM_MELTW_FLAG_UNARY_GS_ROWS            = LIBXSMM_MELTW_FLAG_UNARY_REDUCE_COLS,
+  LIBXSMM_MELTW_FLAG_UNARY_GS_COLS            = LIBXSMM_MELTW_FLAG_UNARY_REDUCE_ROWS,
+  LIBXSMM_MELTW_FLAG_UNARY_GS_OFFS            = 8192,
+  LIBXSMM_MELTW_FLAG_UNARY_NTS_HINT           = 16384
```
[the remark this answers is not in the document]
Comment (alheinecke) [reply include/libxsmm_typedefs.h:248]: yep....

**Comment by alheinecke** [comment]
yes, equations need extra enabling, let's cross this bridge, when we see need.

**Review by egeor — APPROVED** [review]

File: src/generator_mateltwise_sse_avx_avx512.c (hunk 1 above)
Comment (alheinecke) [reply src/generator_mateltwise_sse_avx_avx512.c]: done

**Review by egeor — APPROVED** [review]

**Pull request merged**